### PR TITLE
Refactor of Context instance hierarchy

### PR DESCRIPTION
**Pull request by Odomontois**

### CODE REVIEW

File: core/src/main/scala/tofu/Context.scala [hunk 1]
```diff
@@ -20,48 +36,68 @@ trait Context[F[_]] {
...
-object WithContext extends ContextInstances[WithContext] {
+object WithContext {
   def apply[F[_], C](implicit ctx: WithContext[F, C]): WithContext[F, C] = ctx
-
 }
 
-trait ContextInstances[TC[f[_], r] >: WithContext[f, r]] extends LocalInstances[TC]
-
+/** lightweight version of ApplicativeLocal
+  * consider to use `WithLocal` or `In` for better type inference
```
Comment (oskin1) [inline core/src/main/scala/tofu/Context.scala:78]:
```suggestion
  * consider using `WithLocal` or `In` for better type inference
```
BTW, `type In[C, F[_]] = WithContext[F, C]` doesn't seem to be a proper alternative for `Local`, does it?

File: core/src/main/scala/tofu/package.scala [hunk 2]
```diff
@@ -3,13 +3,15 @@ import cats.{ApplicativeError, MonadError}
...
+  type In[C, F[_]] = WithContext[F, C]
+
   type HasContext[F[_], C] = Context[F] { type Ctx = C }
 
   type HasLocal[F[_], C] = Local[F] { type Ctx = C }
 
   type HasProvide[F[_], G[_], C] = Provide[F] { type Ctx = C; type Lower[A] = G[A] }
 
-  type HasContextRun[F[_], G[_], C] = RunContext[F] { type Ctx = C; type Lower[A] = G[A] }
+  type HasContextRun[F[_], G[_], C] = RunContext[F] { type Lower[A] = G[A]; type Ctx = C; }
```
Comment (catostrophe) [inline core/src/main/scala/tofu/package.scala:14]: Does this swap affect inference anyhow? Why hasn't HasProvide been changed?

File: core/src/main/scala/tofu/package.scala [hunk 3]
```diff
@@ -3,13 +3,15 @@ import cats.{ApplicativeError, MonadError}
 import cats.data.Ior
 
 package object tofu {
+  type In[C, F[_]] = WithContext[F, C]
```
Comment (catostrophe) [inline core/src/main/scala/tofu/package.scala:6]: What's the use?

File: core/src/main/scala/tofu/Context.scala [hunk 4]
```diff
@@ -74,50 +110,44 @@ trait Provide[F[_]] {
...
   def runEquivalent[A](eq: Equivalent[Ctx, A]): WithRun[F, Lower, A] =
     new RunContextEquivalentInstance[F, Lower, Ctx, A](this, eq)
 }
 
-object RunContext extends RunContextInstances[HasContextRun] {
+object RunContext {
   def apply[F[_]](implicit ctx: RunContext[F]): HasContextRun[F, ctx.Lower, ctx.Ctx] = ctx
   type Aux[F[_], G[_], C] = HasContextRun[F, G, C]
+
+  final implicit def readerTContext[F[_]: Applicative, C]: HasContextRun[ReaderT[F, C, *], F, C] =
```
Comment (catostrophe) [inline core/src/main/scala/tofu/Context.scala:139]: So we still can't avoid this even with Kai's approach?

File: core/src/main/scala/tofu/Context.scala [hunk 5]
```diff
@@ -20,48 +36,68 @@ trait Context[F[_]] {
...
     val context: F[C]       = Applicative[F].pure(c)
   }
+
+  def make[F[_]: Functor, A](fa: F[A]): WithContext[F, A] = new WithContext[F, A] {
+    val functor: Functor[F] = implicitly
+    val context: F[A]       = fa
+  }
+
+  /** a mix-in for supplying environment data type companions with useful things */
+  trait Companion[C] {
```
Comment (catostrophe) [inline core/src/main/scala/tofu/Context.scala:54]: Usage examples?

Comment (Odomontois) [reply core/src/main/scala/tofu/Context.scala:54]: - [ ] usage examples

File: core/src/main/scala/tofu/Context.scala (hunk 1 above)
Comment (Odomontois) [reply core/src/main/scala/tofu/Context.scala:78]: Yes, an artifact of hasty copypasta

File: core/src/main/scala/tofu/Context.scala (hunk 4 above)
Comment (Odomontois) [reply core/src/main/scala/tofu/Context.scala:139]: Why should we avoid this?

Comment (Odomontois) [reply core/src/main/scala/tofu/Context.scala:139]: Yes, unfortunately, scala reject to search parent companions for Aux-types with lifted higher-kinded type-members.
So, removing this and `Provide`'s one makes [ContextSuite](https://github.com/TinkoffCreditSystems/tofu/blob/master/core/src/test/scala/tofu/ContextSuite.scala) fail

File: core/src/main/scala/tofu/package.scala (hunk 2 above)
Comment (Odomontois) [reply core/src/main/scala/tofu/package.scala:14]: No, this change is insignificant. Just a scar of searching how to make `ReaderT` instance work inside `ContextBase` companion

File: core/src/main/scala/tofu/Context.scala (hunk 1 above)
Comment (Odomontois) [inline core/src/main/scala/tofu/Context.scala:78]:
```suggestion
  * consider using `WithLocal` for better type inference
```

File: core/src/main/scala/tofu/Context.scala [hunk 6]
```diff
@@ -20,48 +36,68 @@ trait Context[F[_]] {
...
     val functor: Functor[F] = Functor[F]
     val context: F[C]       = Applicative[F].pure(c)
   }
+
+  def make[F[_]: Functor, A](fa: F[A]): WithContext[F, A] = new WithContext[F, A] {
+    val functor: Functor[F] = implicitly
+    val context: F[A]       = fa
+  }
+
+  /** a mix-in for supplying environment data type companions with useful things */
```
Comment (Odomontois) [inline core/src/main/scala/tofu/Context.scala:53]:
```suggestion
  /** a mix-in for supplying environment data type companions with useful things 
     * {{{
     * // @ClassyOptics
     *  case class MyContext(...)
     * object MyContext extends Context.Companion[MyContext]
     * }}} 
     */
```

File: core/src/main/scala/tofu/package.scala (hunk 3 above)
Comment (Odomontois) [reply core/src/main/scala/tofu/package.scala:6]: Just small type alias to write `X In F` instead of `F WithContext X`

**Review by oskin1 — APPROVED** [review]

**Pull request merged**